### PR TITLE
Introduce a more informative cant_generate error

### PR DESCRIPTION
**Pull request by robertoaloi**
This PR attempts to address the issue #183 by introducing a more informative _cant_generate_ error in case of failure of the random instance generation subsystem due to a strict constraint (e.g. via a `?SUCHTHAT` macro). This is particularly useful for test modules defining a big number of constraints, where identifying the root cause of a failing constraint can be tricky. As a trivial example, given the following module, which contains an _impossible_ generator (`gen_2/0`):

```erlang
-module(test).
-include_lib("proper/include/proper.hrl").

gen_1() ->
  ?SUCHTHAT(A, pos_integer(), A > 0).

gen_2() ->
  ?SUCHTHAT(B, pos_integer(), B < 0).

prop() ->
  ?FORALL({A, B}, {gen_1(), gen_2()}, A =:= B).
```

The current output from proper is:

```erl
2> proper:quickcheck(test:prop()).

Error: Couldn't produce an instance that satisfies all strict constraints after 50 tries.
{error,cant_generate}
```

This leaves the user in doubt if the error refers to the first or the second occorrence of `?SUCHTHAT`.

With the proposed patch, the root cause of the issue becomes more obvious:

```erl
Error: Couldn't produce an instance that satisfies all strict constraints from test:'-gen_2/0-fun-0-' after 50 tries.
{error,
    {cant_generate,
        {'$type',
            [{env,{1,inf}},
             {generator,{typed,#Fun<proper_types.13.103056237>}},
             {is_instance,{typed,#Fun<proper_types.14.103056237>}},
             {shrinkers,[#Fun<proper_types.15.103056237>]},
             {kind,basic},
             {constraints,[{#Fun<test.1.106658561>,true}]}]}}}
```

### CODE REVIEW

**Comment by TheGeorge** [comment]
I agree, that the current output is not very verbose, but I think something
like line number and file or module would be even more helpful here.

**Comment by robertoaloi** [comment]
Absolutely. This PR is just to start the discussion on the topic. I totally agree that line number + file would be clearer.

**Comment by robertoaloi** [comment]
@TheGeorge Do you have a suggestion on how to get the line number for the failing constraint? Two alternatives that come to my mind are:

A. If the module is compiled with `debug_info`, one could use the `beam_lib` and `code` API to retrieve the line number. This sounds extremely ugly and fragile, though, and I would not go for it.
B. Modify the `SUCHTHAT*` macros and/or the `proper_types:add_constraint` function so that, when a constraint is added, we store a reference to the place where the constraint is defined (e.g. using the `?MODULE` and `?LINE` macros). This would be cleaner, but it potentially requires a bigger change.|

I'm open to suggestions.

**Comment by TheGeorge** [comment]
I was thinking about options B. You could store this information as type property `info` or similar.

**Comment by kostis** [comment]
First of all, thanks for opening this PR @robertoaloi !

Now, rather than focusing on "implementation details", such as how to get line info, I would like to first have a more general discussion on what is the _PropEr_ thing for this PR to do.

The original issue (#183) complained, rightfully so, that

> When the random instance generation subsystem fails to produce an instance that satisfies some `?SUCHTHAT` constraint, a `cant_generate` error is returned to the user. ... A hint could be provided to the user by analyzing the failing constraints, instead. ... The Type could instead be inspected, the constraints extracted (e.g. via proper_types:get_prop(constraints, Type) and further analyzed to return useful information about the constraints causing the failure.

So, ideally, the information that is returned to the user should _explain_ the reason(s) for failing to generate terms that satisfy the constraints, i.e., which constraints are unsatisfiable and/or conflicting. 
This is not done by the current PR.  Instead, the current PR:

1. Prints an error message which is better than the previous one.
2. Generates (and returns) a quite incomprehensible error term (the second element of the `cant_generate` tuple).

I think that point 1 above is good.  Perhaps it can be slightly improved by _massaging_ the name of the (function containing the) generator so that instead of printing:
```erl
Error: Couldn't produce an instance that satisfies all strict constraints from test:'-gen_2/0-fun-0-' after 50 tries.
```
it prints
```erl
Error: Couldn't produce an instance that satisfies all strict constraints from test:gen_2/0' after 50 tries.
```
Whether to include in the error message the line number of the function or not is a secondary issue, IMO.

As to point 2, I see very little reason to return PropEr's internal representation to the user -- in fact, arguably this is the wrong thing to do because it makes changing this representation more difficult to do in the future. So, this part of the PR needs to change. Ideally, it should provide the list of constraints that are unsatisfiable and/or difficult to satisfy. However, this may require "deep surgery in PropEr's code base" and/or be difficult to do accurately in all cases (running the risk of being more confusing to the user rather than helpful).  So, perhaps the term which is returned can be just:
```erlang
{cant_generate, {M,F,A}}
```
i.e., the MFA of the generator or the MFA where the generator is contained.


One more thing: the PR should contain more tests, perhaps the test in #183 and some others.  Also, if we go with my suggestion for returning the MFA as part of the `cant_generate` tuple, the existing tests of PropEr's test suite that were changed in this PR should be modified to match on the appropriate MFA (rather than matching on an `_`, as they currently do).

**Comment by TheGeorge** [comment]
I agree, that printing the internal representation is not the right thing to do unless we can somehow pretty-print the constraint. 

The problem is that PropEr and it's macros don't work with MFAs but with funs. So returning `{cant_generate, {M, F, A}}` will not work. I think the line number and the file are would be a good starting point for locating which constraint fails. Right now it is very difficult to locate the constraint since next to no information is given.

**Comment by kostis** [comment]
> The problem is that PropEr and it's macros don't work with MFAs but with funs.

Yes, I know that, but funs in Erlang are _always_ associated with the MFA where they are defined.

For example, take a look at the error message that is printed by this PR. The `test:'-gen_2/0-fun-0-'` fun contains all information about the MFA where it is defined.  It's both possible and easy to reconstruct the MFA from the fun's name.

**Comment by TheGeorge** [comment]
Ah nice, then it works with MFA. I still would like to have the line
number. It gives some additional information (e.g. if multiple constraints
are defined in one function) and is just easy to use.

**Comment by robertoaloi** [comment]
Hi again.

I totally agree on dropping the internal representation of the type from the value 'returned' to the user. It was there only to show you what was available and what could potentially be extracted, but I see no point in showing that to the user in the final solution. This is also the reason why I used a wildcard for the term in the tests for the time being.

I like Kostis' suggestion to return a MFA, since it's a good compromise between having no information at all and having the file and line number (which I agree is the ideal end state of this, but may require a lot of more work for little advantage). The reason why I didn't attempt at parsing the fun name is that I didn't know how stable the fun naming schema was in Erlang but, after a little research, it looks like it has been quite stable. There's actually code in OTP which already uses the same strategy to fetch the 'parent', so it looks like it's safe to do what Kostis suggested:

https://github.com/erlang/otp/blob/423f078e3b8603cbd8bfea3be8709092a861a36e/lib/eunit/src/eunit_lib.erl#L379

I will also try to extend the tests a bit. One of Kostis' comments above contains an important keyword:

> So, ideally, the information that is returned to the user should explain the reason(s) for failing to generate terms that satisfy the constraints, i.e., which constraints are unsatisfiable and/or **conflicting**.

I want to see how the new code will behave (i.e. add a test) in the case of 'conflicting' constraints. I also need to check other constraints such as the ones produced by `?SUCHTHATMAYBE`, which may (or may not) reserve some surprise.

**Comment by kostis** [comment]
@robertoaloi : Sounds like a very good plan.  Looking forward to it!

**Comment by robertoaloi** [comment]
Then I'll proceed with the changes.

It looks like the `?SUCHTHATMAYBE` is not an issue since it never results in a constraint failure (by definition). Also, in the case of multiple failing constraints, it looks like we can probably focus only on the first one and ignore the others. Or would you prefer to be safe and return/deal with all the parents from the constraint MFAs?
Finally, whilst we could just the API from `eunit_lib` to fetch the parent, I'm not sure it's worth to create a dependency from proper to eunit just for this. How do you feel about that?

**Comment by kostis** [comment]
> Also, in the case of multiple failing constraints, it looks like we can probably focus only on the first one and ignore the others.

Focus on the first one.  This also makes the return value always having just one MFA instead of an MFA list.

> Finally, whilst we could just the API from eunit_lib to fetch the parent, I'm not sure it's worth to create a dependency from proper to eunit just for this.

I would prefer no dependency. Just copy the code and leave a comment that this comes from `eunit_lib`.

File: test/proper_tests.erl [hunk 1]
```diff
@@ -982,7 +983,7 @@ false_props_test_() ->
      ?_fails(error_statem:prop_simple())].
 
 error_props_test_() ->
-    [?_errorsOut({cant_generate,_},
+    [?_errorsOut({cant_generate,{proper_tests,error_props_test_,0}},
```
Comment (kostis) [inline test/proper_tests.erl]: `?MODULE` should be better here instead of `proper_tests`.
Same for line 1036 below.

**Comment by robertoaloi** [comment]
It looks like I need to amend the code from `eunit_lib` to ensure backward compatibility.

**Comment by kostis** [comment]
The `fun_parent` function exists in previous OTP versions.  For example, in 17.0 its code looks as follows
```erlang
fun_parent(F) ->
    {module, M} = erlang:fun_info(F, module),
    {name, N} = erlang:fun_info(F, name),
    case erlang:fun_info(F, type) of
        {type, external} ->
            {arity, A} = erlang:fun_info(F, arity),
            {M, N, A};
        {type, local} ->
            [$-|S] = atom_to_list(N),
            C1 = string:chr(S, $/),
            C2 = string:chr(S, $-),
            {M, list_to_atom(string:sub_string(S, 1, C1 - 1)),
             list_to_integer(string:sub_string(S, C1 + 1, C2 - 1))}
    end.
```
Perhaps you can factor out the parts that need to differ before and after 20.x and use a code similar to:
```erlang
-ifdef(AT_LEAST_20).
...   % use string:split here
-else.
...
-endif.
```
the `AT_LEAST_20` macro is/should be available to all modules by the `rebar.config` file.

**Comment by kostis** [comment]
On second thought, perhaps calling `eunit_lib:fun_parent` directly will do the trick.

**Comment by robertoaloi** [comment]
I saw your comments only now, after I included my latest changes. I just pushed a version of the `fun_parent/1` function which should work for all recent versions of OTP, even if it's not the most beautiful one.

If I had to pick between the `AT_LEAST_20` macro and the library call I would vote for the library call. Let me know if you are happy with the current function or if you'd like me to instead invoke the library.

**Comment by TheGeorge** [comment]
> > Also, in the case of multiple failing constraints, it looks like we can probably focus only on the first one and ignore the others.

> 
> Focus on the first one. This also makes the return value always having just one MFA instead of an MFA list.

Is this correct though? When generators have multiple constraints, we cannot be sure that the first one is the one that fails. We might report the MFA for the wrong constraint.

**Comment by kostis** [comment]
I am not sure I get the point of your example.  Have you tried it?

My understanding is that we will (and should) return `{cant_generate, {Mod, myint, 0}}` in this case.

**Comment by kostis** [comment]
@robertoaloi wrote
> If I had to pick between the AT_LEAST_20 macro and the library call I would vote for the library call. Let me know if you are happy with the current function or if you'd like me to instead invoke the library.

I gave it a bit of a thought and I think that the chances that `eunit_lib:parent_fun/1` will disappear or do something totally different at some future point are quite slim.  Let's use that instead, rather than duplicating it in a way that is different both from that of Erlang/OTP 17.0 and of 20.0.

When you do that change, please make sure that you also add `eunit` in the list of applications in dialyzer's PLT (this is in the `Makefile` of PropEr).

Once some more tests have been added, I will merge this PR.

**Comment by robertoaloi** [comment]
@TheGeorge The approach works because the `Condition` specified in the `?SUCHTHAT` is wrapped in a `fun` by the macro and that fun is the 'constraint' which gets added to the 'type'.

@kostis Agreed on using the library.

**Comment by robertoaloi** [comment]
Actually, @TheGeorge 's intuition is correct, even if the counter-example provided above is not. See the following one, instead:

```
-module(test).
-include_lib("proper/include/proper.hrl").
-compile(export_all).

possible() ->
  ?SUCHTHAT(X, pos_integer(), X > 0).

impossible() ->
  ?SUCHTHAT(X, possible(), X < 0).

impossible_gen() ->
  ?LET(Impossible, impossible(), Impossible).

impossible_prop() ->
  ?FORALL(_X, impossible_gen(), true).
```

This would return:

```
{error,{cant_generate,{test,possible,0}}}
```

Which is incorrect.

The point is that for some reason I assumed that the boolean contained in the constraint was an indication of a failing constraint, but my assumption was incorrect. That boolean flag simply indicates if the constraint is strict or not. So, I don't think we can return a single MFA.

**Comment by robertoaloi** [comment]
I modified the code, the respective tests and the doc so that a list of candidate MFAs is returned on `cant_generate`. I also added a few extra tests. If you have any other scenario that you would like to test, please advise.

**Comment by TheGeorge** [comment]
Looks good to me

**Comment by kostis** [comment]
@robertoaloi : Thanks again for the PR!

I've reviewed the code and overall it looks good.  The only line that looks a bit weird is the following one in the tests:
```erlang
  ?_errorsOut({cant_generate,[{proper_statem,commands,4}]}, prec_false:prop_simple()),
```
What's happening in this test is that a list of commands cannot be generated because some precondition (defined in the `prec_false` module) is false. Still, the `cant_generate` error which is returned mentions only the `proper_statem:commands/4` generator, which is actually nowhere mentioned in the `prec_false` module.  This may be confusing (to a user).

Can you check whether something can be done for this case too?  Otherwise, I will merge it as is.

**Comment by robertoaloi** [comment]
@kostis I've been thinking about this as well.

Whilst it's true that the `proper_statem:commands/4`generator is not directly referred from the `prec_false` module, the `proper_statem:commands/1` **is** (even if not via its fully qualified name because of the `import`) and that should already provide a precious hint to the user. Compared to the previous implementation, this basically tells the user that the problem is not in one of her generators, but it's related to the generation of command sequences for proper. Understanding that this has to do with _impossible_ preconditions is a bit of a hard leap, but seems a step towards the right direction. Maybe we could simply add a line in the doc (either next to the `cant_generate`doc or in the `statem` section) to explain things a bit better.

In other words, I don't see any obvious solution to improve this other than modifying the `proper_statem` code  or making a special case (read, a horrible hack) for the `proper_statem` module. I'd avise against both approaches. Instead, I vote for merging this PR as is (maybe with a doc addition) and, potentially, improve things further in a separate PR, where we could add some metadata such as line number (and maybe a _message_ in the case of the statem) when adding a constraint to the type.

**Comment by kostis** [comment]
> Instead, I vote for merging this PR as is (maybe with a doc addition) and, potentially, improve things further in a separate PR, where we could add some metadata such as line number (and maybe a message in the case of the statem) when adding a constraint to the type.

Agreed.  Do you want to have the honor to add something appropriate to the docs, or should I press the merge button and handle it myself separately?

**Comment by robertoaloi** [comment]
Let me give it a try.

**Comment by robertoaloi** [comment]
@kostis I found a sweet spot in the `proper_statem` where to add a reference to the error. I didn't feel like adding it to the `proper` module documentation because that's agnostic of the statem. The PR is ready from my perspective. Let me know if you want me to squash the 7 commits into one. Otherwise we are ready to go!

**Comment by kostis** [comment]
Looks good.

No need to squash the commits -- I will hit the GitHub magic button that does this once Travis finishes.
